File: packages/shared-ts/src/telemetry.ts
import {
  LogLevel,
  ObservabilityEvent,
  TelemetryResultStatus,
} from "./generated/contracts";
import type { TelemetryRecord } from "./generated/contracts";

// 与规范文档保持一致的敏感字段集合，统一在 logger 层做兜底脱敏。
const SENSITIVE_KEYS = new Set([
  "token",
  "refresh_token",
  "password",
  "authorization",
  "cookie",
]);

export type TelemetryPayload = Record<string, unknown>;

export interface TelemetryBaseContext {
  trace_id?: string;
  request_id?: string;
  session_id?: string;
  route?: string;
  platform?: string;
  device_id?: string;
  user_uuid?: string;
  role?: string;
  is_authenticated?: boolean;
}

export interface FrontendTelemetryOptions {
  app: "client" | "web";
  appVersion: string;
  env?: string;
  getBaseContext?: () => TelemetryBaseContext;
  sink?: (record: TelemetryRecord) => void | Promise<void>;
}

export interface TrackEventOptions {
  level?: LogLevel;
  payload?: TelemetryPayload;
  latency_ms?: number;
  status?: TelemetryResultStatus;
  error_code?: string;
  error_message?: string;
}

const defaultSink = (record: TelemetryRecord) => {
  const line = JSON.stringify(record);
  if (record.level === LogLevel.Error || record.level === LogLevel.Warn) {
    console.warn(line);
  } else {
    console.info(line);
  }
};

const toSerializable = (value: unknown): unknown => {
  if (value === null || value === undefined) return null;
  if (
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean"
  ) {
    return value;
  }
  try {
    return JSON.parse(JSON.stringify(value));
  } catch {
    return "[unserializable]";
  }
};

export const sanitizePayload = (payload: TelemetryPayload = {}) => {
  const sanitized: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(payload)) {
    sanitized[k] = SENSITIVE_KEYS.has(k.toLowerCase())
      ? "***"
      : toSerializable(v);
  }
  return sanitized;
};

const sourceAppName = (app: "client" | "web") =>
  app === "client" ? "navix-client" : "navix-web";

export const createFrontendTelemetryLogger = (
  options: FrontendTelemetryOptions,
) => {
  const sink = options.sink ?? defaultSink;
  const localContext: TelemetryBaseContext = {
    platform: options.app,
  };

  const createId = () => {
    if (typeof globalThis.crypto?.randomUUID === "function") {
      return globalThis.crypto.randomUUID().replace(/-/g, "");
    }
    return `${Date.now().toString(16)}${Math.random().toString(16).slice(2, 10)}`;
  };

  const resolveContext = (overrides: TelemetryBaseContext = {}) => {
    const base = options.getBaseContext?.() ?? {};
    const merged = { ...base, ...localContext, ...overrides };
    if (!merged.trace_id) {
      merged.trace_id = createId();
    }
    if (!merged.session_id) {
      merged.session_id = createId();
    }
    return merged;
  };

  const setContext = (patch: TelemetryBaseContext) => {
    Object.assign(localContext, patch);
  };

  const withTraceHeaders = (
    headers?: HeadersInit,
    overrides: TelemetryBaseContext = {},
  ): Headers => {
    const resolved = resolveContext(overrides);
    const nextHeaders = new Headers(headers);
    const requestId = createId();
    if (!nextHeaders.has("x-trace-id") && resolved.trace_id) {
      nextHeaders.set("x-trace-id", resolved.trace_id);
    }
    if (!nextHeaders.has("x-request-id")) {
      nextHeaders.set("x-request-id", requestId);
    }
    localContext.request_id = requestId;
    return nextHeaders;
  };

  const captureResponseContext = (headers?: Headers | null) => {
    if (!headers) return;
    const traceId = headers.get("x-trace-id");
    if (traceId && traceId.trim().length > 0) {
      localContext.trace_id = traceId.trim();
    }
    const requestId = headers.get("x-request-id");
    if (requestId && requestId.trim().length > 0) {
      localContext.request_id = requestId.trim();
    }
  };

  const track = async (
    event: string,
    optionsOrPayload: TrackEventOptions | TelemetryPayload = {},
    overrides: TelemetryBaseContext = {},
  ) => {
    const isFlatPayload =
      optionsOrPayload !== null &&
      typeof optionsOrPayload === "object" &&
      !("payload" in optionsOrPayload) &&
      !("level" in optionsOrPayload);
    const opts: TrackEventOptions = isFlatPayload
      ? { payload: optionsOrPayload as TelemetryPayload }
      : optionsOrPayload;
    const resolved = resolveContext(overrides);
    const level = opts.level ?? LogLevel.Info;
    const hasFailure =
      level === LogLevel.Error ||
      level === LogLevel.Fatal ||
      !!opts.error_code ||
      !!opts.error_message;
    const record: TelemetryRecord = {
      schema_version: "1.0.0",
      event_name: event,
      event_id: createId(),
      timestamp: new Date().toISOString(),
      level,
      source: {
        layer: options.app,
        app: sourceAppName(options.app),
        app_version: options.appVersion,
        env: options.env ?? "local",
      },
      actor: {
        user_uuid: resolved.user_uuid ?? null,
        is_authenticated: resolved.is_authenticated ?? !!resolved.user_uuid,
        role: resolved.role ?? null,
      },
      context: {
        session_id: resolved.session_id ?? createId(),
        trace_id: resolved.trace_id ?? "",
        request_id: resolved.request_id ?? null,
        route: resolved.route ?? null,
        platform: resolved.platform ?? options.app,
        device_id: resolved.device_id ?? null,
      },
      metrics: {
        latency_ms: opts.latency_ms ?? null,
      },
      result: {
        status:
          opts.status ??
          (hasFailure
            ? TelemetryResultStatus.Fail
            : TelemetryResultStatus.Success),
        error_code: opts.error_code ?? null,
        error_message: opts.error_message ?? null,
      },
      payload: sanitizePayload(opts.payload ?? {}),
    };
    await Promise.resolve(sink(record));
    return record;
  };

  const logEvent = async (
    event: ObservabilityEvent,
    level: LogLevel,
    payload: TelemetryPayload = {},
    overrides: TelemetryBaseContext = {},
  ) => track(event, { level, payload }, overrides);

  return {
    track,
    logEvent,
    setContext,
    withTraceHeaders,
    captureResponseContext,
  };
};

File: apps/client/src-tauri/src/modules/telemetry.rs
use shared_rs::dto::telemetry::{
    LogLevel, TelemetryActor, TelemetryContext, TelemetryMetrics, TelemetryRecord, TelemetryResult,
    TelemetryResultStatus, TelemetrySource, TelemetrySourceLayer,
};
use std::collections::BTreeMap;
use std::sync::{Mutex, OnceLock};
use tauri_plugin_http::reqwest::Response;
use uuid::Uuid;

const SENSITIVE_KEYS: [&str; 5] = [
    "token",
    "refresh_token",
    "password",
    "authorization",
    "cookie",
];

static TRACE_CONTEXT: OnceLock<Mutex<String>> = OnceLock::new();
static SESSION_CONTEXT: OnceLock<Mutex<String>> = OnceLock::new();

fn trace_context() -> &'static Mutex<String> {
    TRACE_CONTEXT.get_or_init(|| Mutex::new(String::new()))
}

fn session_context() -> &'static Mutex<String> {
    SESSION_CONTEXT.get_or_init(|| Mutex::new(String::new()))
}

fn redact_if_sensitive(key: &str, value: &str) -> serde_json::Value {
    let lowered = key.to_ascii_lowercase();
    if SENSITIVE_KEYS.iter().any(|s| *s == lowered) {
        serde_json::Value::String("***".to_string())
    } else {
        serde_json::Value::String(value.to_string())
    }
}

fn sanitize_payload(payload: BTreeMap<String, String>) -> BTreeMap<String, serde_json::Value> {
    payload
        .into_iter()
        .map(|(k, v)| {
            let masked = redact_if_sensitive(&k, &v);
            (k, masked)
        })
        .collect()
}

pub fn new_id() -> String {
    Uuid::new_v4().as_simple().to_string()
}

pub fn set_trace_id(trace_id: impl Into<String>) {
    if let Ok(mut guard) = trace_context().lock() {
        *guard = trace_id.into();
    }
}

pub fn ensure_trace_id() -> String {
    if let Ok(mut guard) = trace_context().lock() {
        if guard.is_empty() {
            *guard = new_id();
        }
        return guard.clone();
    }
    new_id()
}

pub fn ensure_session_id() -> String {
    if let Ok(mut guard) = session_context().lock() {
        if guard.is_empty() {
            *guard = new_id();
        }
        return guard.clone();
    }
    new_id()
}

pub fn capture_trace_id_from_response(response: &Response) {
    let Some(value) = response.headers().get("x-trace-id") else {
        return;
    };
    let Ok(trace_id) = value.to_str() else {
        return;
    };
    let trimmed = trace_id.trim();
    if !trimmed.is_empty() {
        set_trace_id(trimmed.to_string());
    }
}

pub fn emit_event(
    event: &str,
    level: LogLevel,
    trace_id: &str,
    attrs: BTreeMap<String, String>,
    operation: &str,
) {
    let mut payload = sanitize_payload(attrs);
    payload.insert(
        "operation".to_string(),
        serde_json::Value::String(operation.to_string()),
    );

    let has_failure = matches!(level, LogLevel::Error | LogLevel::Fatal | LogLevel::Warn);
    let record = TelemetryRecord {
        schema_version: "1.0.0".to_string(),
        event_name: event.to_string(),
        event_id: new_id(),
        timestamp: chrono::Utc::now().to_rfc3339(),
        level: level.as_str().to_string(),
        source: TelemetrySource {
            layer: TelemetrySourceLayer::Client.as_str().to_string(),
            app: "navix-client".to_string(),
            app_version: env!("CARGO_PKG_VERSION").to_string(),
            env: std::env::var("NAVIX_ENV")
                .or_else(|_| std::env::var("APP_ENV"))
                .unwrap_or_else(|_| "local".to_string()),
        },
        actor: TelemetryActor {
            user_uuid: None,
            is_authenticated: false,
            role: None,
        },
        context: TelemetryContext {
            session_id: ensure_session_id(),
            trace_id: trace_id.to_string(),
            request_id: None,
            route: None,
            platform: Some("client".to_string()),
            device_id: None,
        },
        metrics: TelemetryMetrics { latency_ms: None },
        result: TelemetryResult {
            status: if has_failure {
                TelemetryResultStatus::Fail.as_str().to_string()
            } else {
                TelemetryResultStatus::Success.as_str().to_string()
            },
            error_code: payload
                .get("code")
                .and_then(|value| value.as_str())
                .map(str::to_string),
            error_message: payload
                .get("message")
                .and_then(|value| value.as_str())
                .map(str::to_string),
        },
        payload: Some(payload),
    };

    match serde_json::to_string(&record) {
        Ok(line) => match level {
            LogLevel::Error | LogLevel::Fatal => log::error!("{line}"),
            LogLevel::Warn => log::warn!("{line}"),
            LogLevel::Debug => log::debug!("{line}"),
            _ => log::info!("{line}"),
        },
        Err(err) => log::error!("failed to serialize client telemetry record: {err}"),
    }
}
